File: RxUI/src/main/java/com/tamsiree/rxui/view/loading/TLoadingManager.java
package com.tamsiree.rxui.view.loading;

import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.view.animation.LinearInterpolator;

class TLoadingManager implements ValueAnimator.AnimatorUpdateListener {

    private final static int MAX_COLOR_CONSTANT_VALUE = 255;
    private final static int ANIMATION_CYCLE_DURATION = 750; //milis
    private TLoadingView loaderView;
    private Paint rectPaint;
    private LinearGradient linearGradient;
    private float progress;
    private ValueAnimator valueAnimator;
    private float widthWeight = TLoadingProfile.MAX_WEIGHT;
    private float heightWeight = TLoadingProfile.MAX_WEIGHT;
    private boolean useGradient = TLoadingProfile.USE_GRADIENT_DEFAULT;
    private int corners = TLoadingProfile.CORNER_DEFAULT;

    public TLoadingManager(TLoadingView view) {
        loaderView = view;
        init();
    }

    private void init() {
        rectPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.FILTER_BITMAP_FLAG);
        loaderView.setRectColor(rectPaint);
        setValueAnimator(0.5f, 1, ObjectAnimator.INFINITE);
    }

    public void onDraw(Canvas canvas) {
        onDraw(canvas, 0, 0, 0, 0);
    }

    public void onDraw(Canvas canvas, float left_pad, float top_pad, float right_pad, float bottom_pad) {
        float margin_height = canvas.getHeight() * (1 - heightWeight) / 2;
        rectPaint.setAlpha((int) (progress * MAX_COLOR_CONSTANT_VALUE));
        if (useGradient) {
            prepareGradient(canvas.getWidth() * widthWeight);
        }
        canvas.drawRoundRect(new RectF(0 + left_pad,
                        margin_height + top_pad,
                        canvas.getWidth() * widthWeight - right_pad,
                        canvas.getHeight() - margin_height - bottom_pad),
                corners, corners,
                rectPaint);
    }

    public void onSizeChanged() {
        linearGradient = null;
        startLoading();
    }

    private void prepareGradient(float width) {
        if (linearGradient == null) {
            linearGradient = new LinearGradient(0, 0, width, 0, rectPaint.getColor(),
                    TLoadingProfile.COLOR_DEFAULT_GRADIENT, Shader.TileMode.MIRROR);
        }
        rectPaint.setShader(linearGradient);
    }

    public void startLoading() {
        if (valueAnimator != null && !loaderView.valueSet()) {
            valueAnimator.cancel();
            init();
            valueAnimator.start();
        }
    }

    public void setHeightWeight(float heightWeight) {
        this.heightWeight = validateWeight(heightWeight);
    }

    public void setWidthWeight(float widthWeight) {
        this.widthWeight = validateWeight(widthWeight);
    }

    public void setUseGradient(boolean useGradient) {
        this.useGradient = useGradient;
    }

    public void setCorners(int corners) {
        this.corners = corners;
    }

    private float validateWeight(float weight) {
        if (weight > TLoadingProfile.MAX_WEIGHT)
            return TLoadingProfile.MAX_WEIGHT;
        if (weight < TLoadingProfile.MIN_WEIGHT)
            return TLoadingProfile.MIN_WEIGHT;
        return weight;
    }

    public void stopLoading() {
        if (valueAnimator != null) {
            valueAnimator.cancel();
            setValueAnimator(progress, 0, 0);
            valueAnimator.start();
        }
    }

    private void setValueAnimator(float begin, float end, int repeatCount) {
        valueAnimator = ValueAnimator.ofFloat(begin, end);
        valueAnimator.setRepeatCount(repeatCount);
        valueAnimator.setDuration(ANIMATION_CYCLE_DURATION);
        valueAnimator.setRepeatMode(ValueAnimator.REVERSE);
        valueAnimator.setInterpolator(new LinearInterpolator());
        valueAnimator.addUpdateListener(this);
    }

    @Override
    public void onAnimationUpdate(ValueAnimator valueAnimator) {
        progress = (float) valueAnimator.getAnimatedValue();
        loaderView.invalidate();
    }

    public void removeAnimatorUpdateListener() {
        if (valueAnimator != null) {
            valueAnimator.removeUpdateListener(this);
            valueAnimator.cancel();
        }
        progress = 0f;
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/loading/TLoadingManager.kt
package com.tamsiree.rxui.view.loading

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.graphics.*
import android.view.animation.LinearInterpolator

internal class TLoadingManager(private val loaderView: TLoadingView) : ValueAnimator.AnimatorUpdateListener {
    private var rectPaint: Paint? = null
    private var linearGradient: LinearGradient? = null
    private var progress = 0f
    private var valueAnimator: ValueAnimator? = null
    private var widthWeight = TLoadingProfile.MAX_WEIGHT
    private var heightWeight = TLoadingProfile.MAX_WEIGHT
    private var useGradient = TLoadingProfile.USE_GRADIENT_DEFAULT
    private var corners = TLoadingProfile.CORNER_DEFAULT
    private fun init() {
        rectPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)
        loaderView.setRectColor(rectPaint!!)
        setValueAnimator(0.5f, 1f, ObjectAnimator.INFINITE)
    }

    @SuppressLint("DrawAllocation")
    @JvmOverloads
    fun onDraw(canvas: Canvas, left_pad: Float = 0f, top_pad: Float = 0f, right_pad: Float = 0f, bottom_pad: Float = 0f) {
        val margin_height = canvas.height * (1 - heightWeight) / 2
        rectPaint!!.alpha = (progress * MAX_COLOR_CONSTANT_VALUE).toInt()
        if (useGradient) {
            prepareGradient(canvas.width * widthWeight)
        }
        canvas.drawRoundRect(RectF(0 + left_pad,
                margin_height + top_pad,
                canvas.width * widthWeight - right_pad,
                canvas.height - margin_height - bottom_pad),
                corners.toFloat(), corners.toFloat(),
                rectPaint!!)
    }

    fun onSizeChanged() {
        linearGradient = null
        startLoading()
    }

    private fun prepareGradient(width: Float) {
        if (linearGradient == null) {
            linearGradient = LinearGradient(0f, 0f, width, 0f, rectPaint!!.color,
                    TLoadingProfile.COLOR_DEFAULT_GRADIENT, Shader.TileMode.MIRROR)
        }
        rectPaint!!.shader = linearGradient
    }

    fun startLoading() {
        if (valueAnimator != null && !loaderView.valueSet()) {
            valueAnimator!!.cancel()
            init()
            valueAnimator!!.start()
        }
    }

    fun setHeightWeight(heightWeight: Float) {
        this.heightWeight = validateWeight(heightWeight)
    }

    fun setWidthWeight(widthWeight: Float) {
        this.widthWeight = validateWeight(widthWeight)
    }

    fun setUseGradient(useGradient: Boolean) {
        this.useGradient = useGradient
    }

    fun setCorners(corners: Int) {
        this.corners = corners
    }

    private fun validateWeight(weight: Float): Float {
        if (weight > TLoadingProfile.MAX_WEIGHT) return TLoadingProfile.MAX_WEIGHT
        return if (weight < TLoadingProfile.MIN_WEIGHT) TLoadingProfile.MIN_WEIGHT else weight
    }

    fun stopLoading() {
        if (valueAnimator != null) {
            valueAnimator!!.cancel()
            setValueAnimator(progress, 0f, 0)
            valueAnimator!!.start()
        }
    }

    private fun setValueAnimator(begin: Float, end: Float, repeatCount: Int) {
        valueAnimator = ValueAnimator.ofFloat(begin, end)
        valueAnimator?.repeatCount = repeatCount
        valueAnimator?.duration = ANIMATION_CYCLE_DURATION.toLong()
        valueAnimator?.repeatMode = ValueAnimator.REVERSE
        valueAnimator?.interpolator = LinearInterpolator()
        valueAnimator?.addUpdateListener(this)
    }

    override fun onAnimationUpdate(valueAnimator: ValueAnimator) {
        progress = valueAnimator.animatedValue as Float
        loaderView.invalidate()
    }

    fun removeAnimatorUpdateListener() {
        if (valueAnimator != null) {
            valueAnimator!!.removeUpdateListener(this)
            valueAnimator!!.cancel()
        }
        progress = 0f
    }

    companion object {
        private const val MAX_COLOR_CONSTANT_VALUE = 255
        private const val ANIMATION_CYCLE_DURATION = 750 //milis
    }

    init {
        init()
    }
}